src/routes/instance.rs

```diff
@@ -199,27 +208,100 @@ pub fn admin_users(
     )))
 }
 
-#[post("/admin/users/<id>/ban")]
-pub fn ban(_admin: Admin, id: i32, rockets: PlumeRocket) -> Result<Flash<Redirect>, ErrorPage> {
-    let u = User::get(&*rockets.conn, id)?;
-    u.delete(&*rockets.conn, &rockets.searcher)?;
+/// A structure to handle forms that are a list of items on which actions are applied.
+///
+/// This is for instance the case of the user list in the administration.
+pub struct MultiAction<T> where T: FromStr {
+    ids: Vec<i32>,
+    action: T,
+}
+
+impl<'f, T> FromForm<'f> for MultiAction<T> where T: FromStr {
+    type Error = ();
+
+    fn from_form(items: &mut FormItems, _strict: bool) -> Result<Self, Self::Error> {
+        let (ids, act) = items.fold((vec![], None), |(mut ids, act), item| {
+            let (name, val) = item.key_value_decoded();
+
+            if name == "action" {
+                (ids, T::from_str(&val).ok())
+            } else if let Ok(id) = name.parse::<i32>() {
+                ids.push(id);
+                (ids, act)
+            } else {
+                (ids, act)
+            }
+        });
+
+        if let Some(act) = act {
+            Ok(MultiAction {
+                ids,
+                action: act,
+            })
+        } else {
+            Err(())
+        }
+    }
+}
+
+pub enum UserActions {
+    Admin,
+    RevokeAdmin,
+    Moderator,
+    RevokeModerator,
+    Ban,
+}
+
+impl FromStr for UserActions {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "admin" => Ok(UserActions::Admin),
+            "un-admin" => Ok(UserActions::RevokeAdmin),
+            "moderator" => Ok(UserActions::Moderator),
+            "un-moderator" => Ok(UserActions::RevokeModerator),
+            "ban" => Ok(UserActions::Ban),
+            _ => Err(()),
+        }
+    }
+}
+
+#[post("/admin/users/edit", data = "<form>")]
+pub fn edit_users(_mod: Moderator, form: LenientForm<MultiAction<UserActions>>, rockets: PlumeRocket) -> Result<Flash<Redirect>, ErrorPage> {
```

Comment:
if I'm reading this right, and admin can't actually edit users, and a moderator can grant themselves admin rights

Reply:
Indeed.

I will try to make permissions an enum as you suggested.

For the UI part, the disadvantage of your proposal is that you can't apply the same action to a lot of users at the same time (but I don't know if it is really needed?).

Reply:
maybe for mass-ban, but probably not for mass-granting-privileges